### DISCUSSION TITLE
Is injecting into self explicitly not allowed?

### DISCUSSION THREAD

**Discussion by lostmsu** [microsoft/Detours#281] [Q&A] [answered]
I am not getting any errors from `DetourUpdateProcessWithDll`, but my DLL is not getting loaded.

I tried using both the handle returned from `GetCurrentProcess()` and explicitly opening own PID using `OpenProcess`.

Even if it is not supported, I think it is a bug that `DetourUpdateProcessWithDll` completes successfully.

**Comment by sylveon** [comment, marked as answer]
DetourUpdateProcessWithDll is for internal use of DetourCreateProcessWithDll; it creates the process suspended, adds the DLL to the process's IAT using DetourUpdateProcessWithDll, then resumes the process, causing the loader to load that DLL. If you apply it to an already running process, then that won't work as the loader has already processed the IAT.

If you want to forcefully load a DLL into a process that has already been launched, try using CreateRemoteThread + LoadLibrary (or just LoadLibrary if it's in your own process).

**Reply by lostmsu** [reply]
Thanks for the explanation.

Sounds like if I want to expose `DetourUpdateProcessWithDll` my wrapper would need to check if the target process is in a suitable state.

Does Detours simplify `CreateRemoteThread` in any way?

**Reply by sylveon** [reply]
No, it doesn't simplify it but it's also very simple to use (error handling and resource management omitted):
```cpp
std::wstring dllLocation = L"C:\\thing.dll";
HANDLE hProcess = OpenProcess(PROCESS_CREATE_THREAD | PROCESS_QUERY_INFORMATION | PROCESS_VM_OPERATION | PROCESS_VM_READ | PROCESS_VM_WRITE, false, pid);

const size_t size = (dllLocation.length() + 1) * sizeof(wchar_t);
void* remotePtr = VirtualAllocEx(hProcess, nullptr, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);

WriteProcessMemory(hProcess, remotePtr, dllLocation.c_str(), size, nullptr);
CreateRemoteThread(hProcess, nullptr, 0, reinterpret_cast<LPTHREAD_START_ROUTINE>(LoadLibrary), remotePtr, 0, nullptr);
```
